### PR TITLE
fix(db): re-file shared-feed ticks onto the climber's own board

### LINKED ISSUE #5121: 🐞 Bug report: Just had a session on my home wall Tranquility. Go to Home, select Tranquility from the top drop down...
Reported from the Boardsesh mobile app.

## Comment

Just had a session on my home wall Tranquility. Go to Home, select Tranquility from the top drop down list and it says there are no recent sessions. It’s always been like that. Sessions don’t show up for some reason.

## Metadata

| Field | Value |
| --- | --- |
| Platform | ios |
| App version | 2.4.0 (13) |
| Source | drawer-bug |
| Board | moonboard / layout 6 / size 1 / sets [24,25,26,27] @ 40° |
| Climb | GME6 (6c/V5) |
| Session | unknown |
| URL | /profile |
| User agent | unknown |
| Feedback row | app_feedback #10237 |

## Contact

✅ Reporter agreed to follow-up — resolve **app_feedback #10237** in the DB to reach them.

### PR DESCRIPTION
**Pull request by marcodejongh**
## Summary

The data half of #5121. The code path is fixed in #5239; this is the separate decision about the rows already written.

Ticks logged on a wall with no BLE serial were filed against the **system-owned per-config shared feed** — one row globally per (board type, layout, size, sets) — instead of the board the climber owns. Their own board then reads as empty everywhere it is scoped by `board_id`: the Home tab's board picker, board stats, board leaderboards. The reporter's 172 MoonBoard ticks are all on feed `261536` and none on their board `770338`.

A code fix alone leaves that history invisible, so this re-files it.

### The rule

A tick moves only when its climber owns **exactly one** non-deleted board with the feed's `(board_type, layout_id, size_id, normalised set_ids)`.

- **Two same-config boards** is #4174's "same wall at home and at the gym". Nothing in the row says which one the climber was standing at, so those are counted and left alone.
- **No matching board** means the shared feed is where the tick belongs — the fixed code still files those there.
- Set ids are compared **normalised**: `createBoard` persists whatever order it was handed, so a board stored `25,26,27,24` is the same wall as a feed keyed `24,25,26,27`.

### Dry run against production (2026-09-07)

```
Found 46 per-config shared feed boards
Found 12653 ticks filed on those feeds

Move       10704 ticks across 650 climbers onto their own board
Ambiguous  180 ticks from 7 climbers who own several boards of that config
No board   1769 ticks whose climber owns no board of that config (the feed is correct)
```

The reporter's case is in there: 172 ticks, `261536` → `770338`.

### Safety

- Snapshot of every `(tick uuid, old board id, new board id)` is written **before** any UPDATE; `--revert <snapshot.json>` restores it.
- The revert only touches rows still holding the board this script wrote, so a later change is never clobbered.
- Every write is guarded on the feed id the plan was built against, so a tick re-filed by the fixed code between the read and the write keeps its newer board.
- One transaction across all batches — no half-applied state to reason about after a crash.
- Safe to re-run: a tick already moved off a feed stops matching.
- The Redis board-stats cache has a 60 s TTL, so board stats self-heal with no extra invalidation.

The selection rule is pure and unit-tested (`backfill-shared-feed-tick-boards-helpers.test.ts`): single match, ambiguous, no match, another climber's board, and a tick that moved off its feed between the two reads.

**Approved by maintainer 2026-09-07.**

Fixes #5121

## Release Notes

- Sessions you logged on your own board before today now show up under that board, instead of staying in a feed shared with every other wall like it.

## Test plan

1. Open Home, pick your board from the top dropdown.
2. Sessions from before today appear under "Recent sessions".
3. Open the board sheet — its stats count those sends.

## Risk

Risk: 4/5 — rewrites `board_id` on ~10.7k tick rows. Snapshot-backed and revertible; no rows are created or deleted, and ambiguous rows are skipped rather than guessed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01724XLTFhu7gkLxBeYw8s5Y

### CODE REVIEW

**Comment by marcodejongh** [comment]
Addressed the review notes after rebasing onto main `f37fff303`:

- Forward and revert share 500-UUID batches inside one atomic transaction, retaining the current-board guard. Real Postgres statement-level fixtures verify batch sizes, concurrent-edit preservation, unchanged tick owners, and rollback after a later batch fails.
- Owner reads are also bounded to 500 IDs and explicitly exclude the system owner.
- `userBoards.setIds` is `text('set_ids').notNull()` in `packages/db/src/schema/app/boards.ts`; the selected value and helper input are non-null strings. A nullable guard would be redundant, so that contract is preserved.
- A mixed-climber, multi-feed fixture covers movable, ambiguous, and missing-board counters, including a size-only mismatch.

Both directions now default to database dry-run; writes require explicit `--apply`. The legacy dry-run flag remains supported, conflicting/unknown options fail before connecting, and help plus the operator guide document the commands. Plans and recovery snapshots use exclusive file creation so a later run cannot overwrite the recovery file; a real filesystem test preserves existing bytes. The historical approval note is explicitly historical and does not authorize a current run.

Validation: 14/14 tests with no skips, full repository check, all 53 typecheck tasks, PR body gates and commit hook pass. Tests used temporary tables on an explicitly local connection and rolled back every fixture. No production connection or repair run was performed. Fresh CI is pending on the draft PR.

**Comment by marcodejongh** [comment]
Fixed the final wording note in e14c5e2bf: the forward dry-run reports “No database writes” and then names the plan file it creates. This is a one-line message correction; full repository check, all 53 typecheck tasks and the commit hook pass. No repair command or production database action was run.

**Comment by marcodejongh** [comment]
Addressed the snapshot-validation note: the complete JSON document, timestamp, entries array, each tick ID and positive safe-integer board IDs are checked before creating a database connection. Corrupt files receive a clear error; the malformed-second-entry regression proves validation cannot start applying an earlier valid row. Generated snapshots and empty plans remain valid.

All 15 pure planning/argument/snapshot tests pass, with full repository check, all 53 typecheck tasks and the mandatory hook. The earlier real-Postgres forward/revert/rollback coverage remains unchanged; its explicit local-database opt-in is documented, and no CI integration run is claimed. Current board IDs fit safe integers; recovery inputs now reject unsafe values. No production command was run.

**Comment by marcodejongh** [comment]
Addressed the latest timestamp minor: `applyMoveBatches` captures one timestamp before its transaction and reuses it for every batch. The existing 1001-row local PostgreSQL regression now asserts one timestamp across forward batches and across reverse batches; current-board guards and untouched newer edits remain verified. All 17 tests pass (15 pure plus 2 local-database tests), with no skipped database cases.

The integration tests intentionally require an explicitly local opt-in connection and roll back temporary fixtures. Their absence from standard CI remains documented in the PR body; no production run is authorized or performed. The CLI helper tests cover default dry-run, invalid/conflicting flags, and recovery snapshot validation before connecting, while the full operator flow still calls for inspecting the generated plan before any explicit apply. The known historical repair size remains appropriate for the existing initial read.

File: packages/db/scripts/backfill-shared-feed-tick-boards.ts [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+      console.log('No per-config shared feed boards found. Nothing to do.');
+      return;
+    }
+    const feedIds = feeds.map((feed) => Number(feed.id));
+    console.log(`Found ${feeds.length} per-config shared feed boards`);
+
+    const ticks = await db
+      .select({ uuid: boardseshTicks.uuid, userId: boardseshTicks.userId, boardId: boardseshTicks.boardId })
+      .from(boardseshTicks)
+      .where(inArray(boardseshTicks.boardId, feedIds));
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/db/scripts/backfill-shared-feed-tick-boards.ts]: Fixed in 7221eb54bae7f71b3870ad5639c06d389f892371. The actual loader now joins each tick’s session and active session wall. Planning follows saveTick’s full-config session-first rule: a foreign party wall wins over zero, one, or multiple owned matches; a valid session already on the feed stays there. Missing/deleted/mismatched sessions fall back to the existing unique-owned-wall rule. Ownership and session IDs are unchanged.

All 21 tests passed, including 3 real local-Postgres fixtures. The joined-loader fixture applies and reverts 8 moves and verifies unchanged owners/session IDs; both original bounded/CAS/rollback cases still pass. Full checks, all 53 typecheck tasks and the commit hook passed. No production count, plan or repair was executed. The PR stays draft for fresh CI.

**Comment by marcodejongh** [comment]
Addressed all four follow-up items in a57534b43f15a0332a3e116a0ee541e840f11ece. The runbook now explains that touched rows stay locked until the whole operation commits/rolls back and recommends a low-traffic window for separately authorized writes. Atomic batching is unchanged. Default filenames include plan/apply plus an ISO timestamp; explicit existing paths are checked before database queries, while exclusive-create remains the concurrent-write guard.

The existing backend CI job now runs all three Node suites on its first shard using its disposable loopback PostgreSQL service. CI fails for a missing URL, test failure, skipped integration cases or a zero-test report; remote URLs are refused before connecting. No new service or production credential is involved.

All 22 repair tests (including 3 real local-Postgres cases), 57 existing workflow tests, full repository checks, all 53 typecheck tasks and the mandatory hook passed. The exact TAP guard rejected failed/skipped/empty/zero-test fixtures. CLI negative controls verified the early collision error preserves recovery bytes before database use. No production query or repair was executed. Draft remains for fresh CI.

**Comment by marcodejongh** [comment]
Checked the ID type against the actual schema: [userBoards.id](https://github.com/boardsesh/boardsesh/blob/a57534b43f15a0332a3e116a0ee541e840f11ece/packages/db/src/schema/app/boards.ts#L23) uses `bigserial` with `mode: 'number'`. Drizzle therefore returns the selected `sessionBoard.id` as a number. The real PostgreSQL integration fixture declares this column as bigint, executes `loadSharedFeedTicks`, and asserts `sessionRetained === 1`; that test passed locally and is now mandatory in the first backend CI shard. No extra coercion is needed.

The TAP check deliberately requires an explicit zero-skips footer from the pinned Node reporter, as well as positive test count and all tests passing. Its recorded controls reject empty, failed, skipped and zero-test output. If the reporter changes, failing the gate is preferable to silently losing skip accounting; the guard can then be updated alongside the reporter with those same controls. This does not require a speculative code change now.

File: docs/shared-feed-tick-repair.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+The default is a database dry-run. It still writes a local JSON plan; keep that file for review. Every plan and recovery snapshot is created exclusively: an existing file causes the command to fail before applying any database changes. Existing explicit paths are rejected before database queries; the exclusive create still guards a path created concurrently after that check. Defaults include `plan` or `apply` and a filesystem-safe ISO timestamp, so a same-day preview and apply use different paths. Choose a new `--out` path when supplying one explicitly. The legacy `--dry-run` flag remains supported. Unknown options, missing filenames and conflicting `--apply --dry-run` flags are rejected before connecting.
+
+```sh
+vp run db:backfill-shared-feed-tick-boards -- --out shared-feed-plan.json
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply docs/shared-feed-tick-repair.md:8]: Fixed in f4367f25ee7b4e405721569627790692860f3674. `vite.config.ts` now registers an uncached root `db:backfill-shared-feed-tick-boards` task that delegates to the package script. `vp run db:backfill-shared-feed-tick-boards -- --help` reached the script help path before `createScriptDb`; no database was opened.

File: packages/db/scripts/backfill-shared-feed-tick-boards.ts [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+    for (const { fromBoardId, toBoardId, uuids } of groups.values()) {
+      for (let offset = 0; offset < uuids.length; offset += BATCH_SIZE) {
+        const rows = await transaction
+          .update(boardseshTicks)
+          .set({ boardId: toBoardId, updatedAt })
+          .where(
+            and(
+              inArray(boardseshTicks.uuid, uuids.slice(offset, offset + BATCH_SIZE)),
+              eq(boardseshTicks.boardId, fromBoardId),
+            ),
```
[the remark this answers is not in the document]
Comment (marcodejongh) [reply packages/db/scripts/backfill-shared-feed-tick-boards.ts]: Fixed in f4367f25ee7b4e405721569627790692860f3674. Forward apply now uses Drizzle serializable transaction options to reload planned source ticks, session walls, shared-feed configuration, and live owned boards in 500-ID chunks, then rebuilds the pure plan. A changed or ambiguous exact destination aborts the entire write; already-refiled rows retain their compare-and-set skip and no replacement target is chosen. Real local PostgreSQL coverage exercises deleted/reconfigured/re-owned destinations, a second owned board, a changed session, source feed reconfiguration, and a late second batch rollback.

**Comment by marcodejongh** [comment]
Addressed in `5194730fb1e4ea1e787a4b425596e156ae6f01a1`: shared-feed tick reads now use 500 distinct feed IDs per query, including duplicate-input protection, and `main()` reuses `loadOwnedBoards` instead of repeating its query and normalization. The session joins and forward/revert transaction guards are unchanged. A real temporary-table test spans 1,001 feeds, a repeated feed ID and an unrelated tick. All 31 helper/CLI/local PostgreSQL tests passed without skips; all 53 typecheck tasks, lint/format, hook and independent paired review passed.

The current Node 22.22.2 TAP run emits `# tests 31`, `# pass 31` and `# skipped 0`. The CI guard deliberately requires a complete no-skip summary; missing summary output fails closed. The reported marker-omission scenario is not reproduced with the supported runner. No production database was accessed.
